test/denormalized_has_many_test.rb

```diff
@@ -141,6 +141,55 @@ def test_deprecated_set_inverse_association_on_cache_hit
     assert_equal item.object_id, associated_record.item.object_id
   end
 
+  def test_returned_records_should_be_readonly_on_cache_hit
+    IdentityCache.with_fetch_read_only_records do
+      Item.fetch(@record.id) # warm cache
+
+      record_from_cache_hit = Item.fetch(@record.id)
+      record_from_cache_hit.fetch_associated_records.each do |record|
+        assert record.readonly?, "Record #{record} wasn't readonly"
+      end
+    end
+  end
+
+  def test_returned_records_should_be_readonly_on_cache_miss
+    IdentityCache.with_fetch_read_only_records do
+      record_from_cache_miss = Item.fetch(@record.id)
+
+      record_from_cache_miss.fetch_associated_records.each do |record|
+        assert record.readonly?, "Record #{record} wasn't readonly"
+      end
+    end
+  end
+
+  def test_db_returned_records_should_never_be_readonly
+    IdentityCache.with_fetch_read_only_records do
+      record_from_db = Item.find(@record.id)
+
+      uncached_records = record_from_db.associated_records
+      uncached_records.each do |record|
+        assert !record.readonly?, "Record #{record} was readonly"
```

Comment:
The intention here was to test the before and after results of `record_from_db.fetch_associated_records` since with the previous `load_and_readonlyify`, `uncached_records.readonly` became true. It asserts that what you mentionned in https://github.com/Shopify/identity_cache/pull/282/files/5647e89b96ff6edcb52a6e3c5a2fd759fb875322#r72793608 is not happening anymore.

Specific tests for cache_hit and miss behavior when read-only is enabled are here: https://github.com/Shopify/identity_cache/blob/df2f676fbccdacb0adbce77709b944e807ed7d36/test/denormalized_has_many_test.rb#L144-L163